### PR TITLE
AArch64: Correctly handle ificmpeq/ificmpne nodes with many global registers

### PR DESCRIPTION
**Pull request by Akira1Saitoh**
Correctly count the integer or address child nodes added to GlRegDeps node
to avoid adding too many integer registers to the register dependencies.
Add a src register of compare branch instructions to register dependencies
so that a src register is assigned when other registers in the register dependencies
are assigned.

Signed-off-by: Akira Saitoh <saiaki@jp.ibm.com>

### CODE REVIEW

**Comment by Akira1Saitoh** [comment]
This will resolve https://github.com/eclipse-openj9/openj9/issues/13502

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 1]
```diff
@@ -143,7 +194,7 @@ if (secondChildNeedsRelocation)
              * We need to use a b.cond instruction instead for that case.
              */
             && ((node->getNumChildren() != 3) ||
-             (node->getChild(2)->getNumChildren() != cg->getLinkage()->getProperties().getNumAllocatableIntegerRegisters())))
+             (getNumberOfIntegerNodesInGlRegDeps(node->getChild(2), cg) < cg->getLinkage()->getProperties().getNumAllocatableIntegerRegisters())))
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ControlFlowEvaluator.cpp:197]: If `src1Reg` is not part of the GlRegDeps, does it need to be accounted for in this check if you will be eventually adding it to the reg dep below?

**Review by 0xdaryl — CHANGES_REQUESTED** [review]

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp (hunk 1 above)
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/ControlFlowEvaluator.cpp:197]: When this check passes, the number of integer child nodes of the GlRegDeps is less than the maximum allocatable integer registers (25 in our case), so even if `src1Reg` is not part of the GlRegDeps, the number of integer registers added to the reg dep does not exceed the limit (25).
If `src1Reg` is part of the GlRegDeps, the number of integer children can be less than or equal to the limit, but to check if `src1Reg` is part of the GlRegDeps, we need to evaluate the GlRegDeps node first.

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 2]
```diff
@@ -95,6 +95,57 @@ TR::Register *OMR::ARM64::TreeEvaluator::gotoEvaluator(TR::Node *node, TR::CodeG
...
+          *      iRegLoad x14
+          *      dRegLoad d31
+          *      dRegLoad d30
+          *
+          */
+         if (!child->getDataType().isFloatingPoint())
+            {
+            numIntNodes++;
+            }
+         }
```
[the remark this answers is not in the document]
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/ControlFlowEvaluator.cpp:134]: Thanks for the suggestion. I changed the body of the loop.

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 3]
```diff
@@ -95,6 +95,57 @@ TR::Register *OMR::ARM64::TreeEvaluator::gotoEvaluator(TR::Node *node, TR::CodeG
...
    }
 
+/**
+ * @brief Returns the number of integer or address child nodes of the GlRegDeps node
+ *
+ * @param[in]  glRegDepsNode: GlRegDeps node
+ * @param[in]             cg: Code Generator
+ * @return the number of integer or address child nodes of the GlRegDeps node
+ */
+static uint32_t getNumberOfIntegerNodesInGlRegDeps(TR::Node *glRegDepsNode, TR::CodeGenerator *cg)
```
[the remark this answers is not in the document]
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: Changed the name.

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 4]
```diff
@@ -95,6 +95,57 @@ TR::Register *OMR::ARM64::TreeEvaluator::gotoEvaluator(TR::Node *node, TR::CodeG
...
+      TR::Node *child = glRegDepsNode->getChild(i);
+      if (child->getOpCodeValue() == TR::PassThrough)
+         {
+         /*
+          *  For PassThrough node, we need to check the data type of the child node of PassThrough.
+          *    GlRegDeps
+          *       PassThrough x15
+          *         ==>acalli
+          */
+         if (!child->getFirstChild()->getDataType().isFloatingPoint())
```
[the remark this answers is not in the document]
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: Thanks for pointing it out. I forgot that vector was supported :-)

**Comment by 0xdaryl** [comment]
Jenkins build aarch64

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**